src/position.ts

```diff
@@ -0,0 +1,305 @@
+import {
+    AA_PATTERN, AA_CODES, PREFIX_CLASS, Prefix,
+} from './constants';
+
+const CDSLIKE_PATT = /(?<pos>-?(\d+|\?))?(?<offset>[-+](\d+|\?))?/;
+const CLASS_FIELD = '@class';
+const PATTERNS = {
+    y: /(?<arm>[pq])((?<majorBand>\d+|\?)(\.(?<minorBand>\d+|\?))?)?/,
+    p: new RegExp(`(?<refAA>${AA_PATTERN})?(?<pos>\\d+|\\?)`),
+    c: CDSLIKE_PATT,
+    n: CDSLIKE_PATT,
+    r: CDSLIKE_PATT,
+} as const;
+
+interface Position {
+    '@class': string;
+    prefix: Prefix;
+}
+
+interface BasicPosition extends Position {
+    pos: number | null;
+}
+
+interface CdsLikePosition extends BasicPosition {
+    prefix: 'c' | 'n' | 'r';
+    offset?: number | null;
+}
+
+interface CytobandPosition extends Position {
+    prefix: 'y';
+    arm: 'p' | 'q';
+    majorBand?: number | null;
+    minorBand?: number | null;
+}
+
+interface ProteinPosition extends BasicPosition {
+    prefix: 'p';
+    refAA: string | null;
+    longRefAA?: string | null;
+}
+
+type AnyPosition = BasicPosition | CytobandPosition | ProteinPosition;
+
+type PrefixMap<P extends Prefix> = (
+    P extends 'y' ? CytobandPosition :
+        P extends 'p' ? ProteinPosition :
+            P extends 'c' | 'r' | 'n' ? CdsLikePosition :
+                P extends 'g' | 'i' | 'e' ? BasicPosition :
+                    AnyPosition
+);
+
+const convertPositionToJson = (position: AnyPosition, exclude = ['prefix', 'longRefAA']) => {
+    const json = {};
+
+    for (const [attr, val] of Object.entries(position)) {
+        if (val !== undefined && !exclude.includes(attr)) {
+            json[attr] = val;
+        }
+    }
+    return json;
+};
+
+const createCytoBandPosition = ({ arm, majorBand, minorBand }): CytobandPosition => {
+    const prefix: Prefix = 'y';
+    const result = {
+        arm, majorBand, minorBand, [CLASS_FIELD]: PREFIX_CLASS[prefix], prefix,
+    };
+
+    if (arm !== 'p' && arm !== 'q') {
+        throw new InputValidationError({
+            message: `cytoband arm must be p or q (${arm})`,
+            violatedAttr: 'arm',
+        });
+    }
+    if (majorBand !== undefined && majorBand !== null) {
+        result.majorBand = Number(majorBand);
+
+        if (Number.isNaN(Number(result.majorBand)) || result.majorBand <= 0) {
+            throw new InputValidationError({
+                message: `majorBand must be a positive integer (${majorBand})`,
+                violatedAttr: 'majorBand',
+            });
+        }
+    }
+    if (minorBand !== undefined && minorBand !== null) {
+        result.minorBand = Number(minorBand);
+
+        if (Number.isNaN(Number(result.minorBand)) || result.minorBand <= 0) {
+            throw new InputValidationError({
+                message: `minorBand must be a positive integer (${minorBand})`,
+                violatedAttr: 'minorBand',
+            });
+        }
+    }
+    return result;
+};
+
+const checkBasicPosition = (pos: any, allowNegative = false): number | null => {
```

Comment:
is it possible to use something other than 'any' here? github actions hates this

Reply:
you could likely use `unknown` which is better

Reply:
I fixed a few of these but some of them are quite complicated and so i will leave them for a future PR